File: examples/threads/SharedQueue.java
import java.util.ArrayList;

/*
 * A simple shared queue that is synchronized for use by multiple threads.
 * @author amit
 */

public class SharedQueue
{
    private ArrayList<String> queue;
    private int maxSize;

    public SharedQueue(int maxSize) {
        queue = new ArrayList<String>();
        this.maxSize = maxSize;
    }


    public synchronized int size() {
        return queue.size();
    }


    public synchronized void putMessage(String msg) throws InterruptedException {
        while (this.size() == maxSize)
            wait();
        queue.add(msg);
        notifyAll(); // for multiple consumers
    }


    public synchronized String getMessage() throws InterruptedException {
        notifyAll(); // use notify for single producer
        while (this.size() == 0)
            wait();
        String message = queue.remove(0); // first element
        queue.remove(message);
        return message;
    }

}

File: examples/golang/threads/SharedQueue.go
//Author: Amit Jain & Paul Ellis
package main

import (
	"sync"
)

type SharedQueue struct{
	queue []string
	maxSize int
	mu sync.Mutex
}

func NewSharedQueue(maxSize int) *SharedQueue{
	//When making the new slice for strings we provide the size
	q := SharedQueue{queue:make([]string, maxSize), maxSize:maxSize}
	return &q
}

func (q *SharedQueue)Size() int{
	var s int
	q.mu.Lock()
	s = len(q.queue)
	q.mu.Unlock()
	return s
}

func (q *SharedQueue)PutMessage(msg string){
	//Here we are waiting for the queue to have space for a new message
	for{
		if q.Size() != q.maxSize {
			break
		}
	}
	q.mu.Lock()
	q.queue = append(q.queue, msg)
	q.mu.Unlock()
}

func (q *SharedQueue)GetMessage() string{
	//We will wait for there to actually be a message in the queue
	for{
		if q.Size() != 0{
			break
		}
	}
	var msg string
	q.mu.Lock()
	msg = q.queue[0]
	//this sets the queue to be the queue from index 1 to the end
	//effectively removing the head from the queue
	q.queue = q.queue[1:]
	q.mu.Unlock()
	return msg
}